src/runtime.js

```diff
@@ -333,6 +346,108 @@ var Runtime = {
     }
     Runtime.loadedDynamicLibraries.push(libModule);
   },
+
+#if BINARYEN
+  // Loads a side module from binary data
+  loadWebAssemblyModule: function(wasm) {
+    var int32View = new Uint32Array(new Uint8Array(wasm.subarray(0, 24)).buffer);
+    assert(int32View[0] == 0x6d736100, 'need to see wasm magic number'); // \0wasm
+    // we should see the dylink section right after the magic number and wasm version
+    assert(wasm[8] === 0, 'need the dylink section to be first')
+    var next = 9;
+    function getLEB() {
+      var ret = 0;
+      var mul = 1;
+      while (1) {
+        var byte = wasm[next++];
+        ret += ((byte & 127) * mul);
+        mul *= 128;
+        if (!(byte & 128)) break;
+      }
+      return ret;
+    }
+    var sectionSize = getLEB();
+    assert(wasm[next] === 6);                 next++; // size of "dylink" string
+    assert(wasm[next] === 'd'.charCodeAt(0)); next++;
+    assert(wasm[next] === 'y'.charCodeAt(0)); next++;
+    assert(wasm[next] === 'l'.charCodeAt(0)); next++;
+    assert(wasm[next] === 'i'.charCodeAt(0)); next++;
+    assert(wasm[next] === 'n'.charCodeAt(0)); next++;
+    assert(wasm[next] === 'k'.charCodeAt(0)); next++;
+    var memorySize = getLEB();
+    var tableSize = getLEB();
+    var env = Module['asmLibraryArg'];
+    // TODO: use only memoryBase and tableBase, need to update asm.js backend
+    var table = Module['wasmTable'];
+    var oldTableSize = table.length;
+    env['memoryBase'] = env['gb'] = Runtime.alignMemory(getMemory(memorySize + Runtime.STACK_ALIGN), Runtime.STACK_ALIGN); // TODO: add to cleanups
+    env['tableBase'] = env['fb'] = oldTableSize;
+    var originalTable = table;
+    table.grow(tableSize);
+    assert(table === originalTable);
+    // zero-initialize memory and table TODO: in some cases we can tell it is already zero initialized
+    for (var i = env['memoryBase']; i < env['memoryBase'] + memorySize; i++) {
+      HEAP8[i] = 0;
+    }
+    for (var i = env['tableBase']; i < env['tableBase'] + tableSize; i++) {
+      table.set(i, null);
+    }
+    // copy currently exported symbols so the new module can import them
+    for (var x in Module) {
+      if (!(x in env)) {
+        env[x] = Module[x];
+      }
+    }
+    var info = {
+      global: Module['asmGlobalArg'],
+      env: env
+    };
+#if ASSERTIONS
+    var oldTable = [];
+    for (var i = 0; i < oldTableSize; i++) {
+      oldTable.push(table.get(i));
+    }
+#endif
+    // create a module from the instance
+    wasm = new WebAssembly.Instance(new WebAssembly.Module(wasm), info);
```

Comment:
Heh, good point. Updating to binary for the binary and instance for the instance.